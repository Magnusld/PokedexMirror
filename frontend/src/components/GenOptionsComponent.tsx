import React from 'react';
import '../style/App.css';
import {Form} from "react-bootstrap";
import {Generation} from "../types";

export function GenOptionsComponent(props: {
  selectedGens: Generation[]
  setSelectedGens: (gens: Generation[]) => void
}) {
  const selectedGens = props.selectedGens
  const setSelectedGens = props.setSelectedGens

  return (
      <div>

      </div>
  )
}