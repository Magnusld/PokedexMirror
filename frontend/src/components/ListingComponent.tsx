import React, {useEffect, useState} from 'react';
import '../style/App.css';
import {PokemonSimple} from "../types";
import {ReactComponent as Bug} from "../assets/typeIconsSvg/Pokémon_Bug_Type_Icon.svg";
import {ReactComponent as Dark} from "../assets/typeIconsSvg/Pokémon_Dark_Type_Icon.svg";
import {ReactComponent as Dragon} from "../assets/typeIconsSvg/Pokémon_Dragon_Type_Icon.svg";
import {ReactComponent as Electric} from "../assets/typeIconsSvg/Pokémon_Electric_Type_Icon.svg";
import {ReactComponent as Fairy} from "../assets/typeIconsSvg/Pokémon_Fairy_Type_Icon.svg";
import {ReactComponent as Fighting} from "../assets/typeIconsSvg/Pokémon_Fighting_Type_Icon.svg";
import {ReactComponent as Fire} from "../assets/typeIconsSvg/Pokémon_Fire_Type_Icon.svg";
import {ReactComponent as Flying} from "../assets/typeIconsSvg/Pokémon_Flying_Type_Icon.svg";
import {ReactComponent as Ghost} from "../assets/typeIconsSvg/Pokémon_Ghost_Type_Icon.svg";
import {ReactComponent as Grass} from "../assets/typeIconsSvg/Pokémon_Grass_Type_Icon.svg";
import {ReactComponent as Ground} from "../assets/typeIconsSvg/Pokémon_Ground_Type_Icon.svg";
import {ReactComponent as Ice} from "../assets/typeIconsSvg/Pokémon_Ice_Type_Icon.svg";
import {ReactComponent as Normal} from "../assets/typeIconsSvg/Pokémon_Normal_Type_Icon.svg";
import {ReactComponent as Poison} from "../assets/typeIconsSvg/Pokémon_Poison_Type_Icon.svg";
import {ReactComponent as Psychic} from "../assets/typeIconsSvg/Pokémon_Psychic_Type_Icon.svg";
import {ReactComponent as Rock} from "../assets/typeIconsSvg/Pokémon_Rock_Type_Icon.svg";
import {ReactComponent as Steel} from "../assets/typeIconsSvg/Pokémon_Steel_Type_Icon.svg";
import {ReactComponent as Water} from "../assets/typeIconsSvg/Pokémon_Water_Type_Icon.svg";
import {getFullSize, getSprite, getThumbnail} from "../util/imageResolver";

export function ListingComponent(props: {
  asGrid: boolean
  pokemon: PokemonSimple
}) {
  const asGrid = props.asGrid
  const pokemon = props.pokemon

  useEffect(() => {
    console.log("Im alive")
  })

  function log() {
    console.log(asGrid)
    console.log("listing"+chooseClassName())
    console.log(getThumbnail(pokemon.pokedexNr))
  }


  /**
   * takes potential type 1 and type 2 and makes it into a list
   * @param types rest param with one or two types
   * @returns 
   */
  function typesToList(...types : string[]) {
    const typeList = types.filter((type) => type.length > 0)
    return typeList;
  }

  let showTypeEmblem = typesToList(pokemon.type1, pokemon.type2).map((Type) => (
      <div>
        {Type === "Bug" ? <Bug className={"typeIcon"}/> : null}
        {Type === "Dark" ? <Dark className={"typeIcon"}/> : null}
        {Type === "Dragon" ? <Dragon className={"typeIcon"}/> : null}
        {Type === "Electric" ? <Electric className={"typeIcon"}/> : null}
        {Type === "Fairy" ? <Fairy className={"typeIcon"}/> : null}
        {Type === "Fighting" ? <Fighting className={"typeIcon"}/> : null}
        {Type === "Fire" ? <Fire className={"typeIcon"}/> : null}
        {Type === "Flying" ? <Flying className={"typeIcon"}/> : null}
        {Type === "Ghost" ? <Ghost className={"typeIcon"}/> : null}
        {Type === "Grass" ? <Grass className={"typeIcon"}/> : null}
        {Type === "Ground" ? <Ground className={"typeIcon"}/> : null}
        {Type === "Ice" ? <Ice className={"typeIcon"}/> : null}
        {Type === "Normal" ? <Normal className={"typeIcon"}/> : null}
        {Type === "Poison" ? <Poison className={"typeIcon"}/> : null}
        {Type === "Psychic" ? <Psychic className={"typeIcon"}/> : null}
        {Type === "Rock" ? <Rock className={"typeIcon"}/> : null}
        {Type === "Steel" ? <Steel className={"typeIcon"}/> : null}
        {Type === "Water" ? <Water className={"typeIcon"}/> : null}
      </div>
  ))

  function chooseClassName() {
    if (asGrid) {
      return "AsGrid"
    } else { return "AsList"}
  }

  function ImageGridStyle() {
    return(
        {backgroundImage: asGrid ? `url(${getFullSize(pokemon.pokedexNr)})` : undefined,
        backgroundRepeat: "no-repeat",
        backgroundPosition: "center 25%",
        backgroundSize: "150px 150px"}
    )
  }

  return (
      <div className={chooseClassName()}
        style={asGrid ? ImageGridStyle() : undefined}>
        <div className={"listing"+chooseClassName()}>
          <div className={"pokedexNumber"+chooseClassName()}>
            <h3>#{pokemon.pokedexNr}</h3>
            {!asGrid && <img className={"ListSprite"} src={getSprite(pokemon.pokedexNr)} alt={pokemon.name}/>}
          </div>
          <div className={"listingInfo"+chooseClassName()}>
            <div>
              <h5 onClick={log}>{pokemon.name}</h5>
            </div>
            <p>Gen {pokemon.generation}</p>
          </div>
          <div className={"typeEmblem"+chooseClassName()}>
            {showTypeEmblem}
          </div>
        </div>
      </div>
  )
}