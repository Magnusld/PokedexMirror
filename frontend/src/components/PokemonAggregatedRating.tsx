import {RatingView} from "react-simple-star-rating";

export type PokemonAggregatedRatingProps = {
    rating: number
}

export default function PokemonAggregatedRating( props: PokemonAggregatedRatingProps) {

    return (
        <div className={"aggregated-rating-container"}>
            <div className={"rating-number"}>Andres vurdering:</div>
            <RatingView ratingValue={props.rating} stars={5} />
        </div>
    )
}