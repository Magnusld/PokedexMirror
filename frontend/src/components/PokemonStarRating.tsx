import {SetStateAction, useEffect, useState} from "react";
import {Rating} from "react-simple-star-rating";
import {gql, useMutation, useQuery} from "@apollo/client";
import {getGuid} from "../util/guidHelper";
import {GET_POKEMON_INFO} from "../pages/Info";

export type PokemonStarRatingProps = {
    pokemonId: number
}

interface RatingDetails {
    "pokemonRating": {
        "rating": number
    } | null
}

const GET_RATING = gql`query Query($where: PokemonRatingWhereUniqueInput!) {
    pokemonRating(where: $where) {
        rating
    }
}`

interface RatingQueryVars {
    "where": {
        "userGuid_pokemonId": {
            "userGuid": string,
            "pokemonId": number
        }
    }
}

const NEW_RATING = gql`mutation CreateRatingMutation($data: RatingCreateInput!) {
    CreateRating(data: $data) {
        id
    }
}`

interface RatingNewDetails {
    "data": {
        "pokemonId": number,
        "userGuid": string,
        "rating": number
    }
}

const SAVE_RATING = gql`mutation ChangeRatingMutation($data: RatingUpdateInput!) {
    ChangeRating(data: $data) {
        id
    }
}`

interface RatingSaveDetails {
    "data": {
        "ratingToUpdate": {
            "pokemonId": number,
            "userGuid": string
        },
        "newRating": number
    }
}

interface UpsertResponse {
    "CreateRating": {
        "id": number
    }
}


export default function PokemonStarRating(props: PokemonStarRatingProps) {
    const [rating, setRating] = useState(0)
    const [hasRated, setHasRated] = useState(false)
    const {data} = useQuery<RatingDetails,RatingQueryVars>(GET_RATING, {
        variables: {where: {userGuid_pokemonId: { userGuid: getGuid(), pokemonId: props.pokemonId}}}
    })
    const [newRatingFunction] = useMutation<UpsertResponse, RatingNewDetails>(NEW_RATING, {
        variables: { data: { pokemonId: props.pokemonId, rating: rating, userGuid: getGuid()}},
        refetchQueries: [GET_RATING, GET_POKEMON_INFO]
    })
    const [saveRatingFunction] = useMutation<UpsertResponse,RatingSaveDetails>(SAVE_RATING, {
        variables: { data: {
            newRating: rating,
                ratingToUpdate: { pokemonId: props.pokemonId, userGuid: getGuid()
                }
            }},
        refetchQueries: [GET_RATING, GET_POKEMON_INFO]
    })

    const handleRating = async (rate: SetStateAction<number>) => {
        await setRating(rate)
        if (!hasRated) {
            await newRatingFunction()
        } else {
            await saveRatingFunction()
        }
    }

    useEffect( () => {
        if (data?.pokemonRating?.rating) {
            setRating(data.pokemonRating.rating)
            setHasRated(true)
        }
    },[data])

    return (
        <div className={"personal-rating-container"}>
            <div className={"rating-number"}>Din vurdering:</div>
            <Rating onClick={handleRating} ratingValue={rating} stars={5} />
        </div>
    )
}