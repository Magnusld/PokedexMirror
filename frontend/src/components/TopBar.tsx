import React from 'react';
import '../style/App.css';
import {Accordion, Button, Dropdown, Form} from "react-bootstrap";
import {ReactComponent as List } from '../assets/list_black_24dp.svg'
import {ReactComponent as Grid } from '../assets/apps_black_24dp.svg';
import {ReactComponent as Search} from "../assets/search_black_24dp.svg";

export function TopBar(props: {
  asGrid: boolean
  setAsGrid: (value: boolean) => void
  setShowGenSelection: (value: boolean) => void
  showGenSelection: boolean
  setShowTypeSelection: (value: boolean) => void
  showTypeSelection: boolean

}) {
  const searchText = ""
  const asGrid = props.asGrid
  const showGenSelection = props.showGenSelection
  const showTypeSelection = props.showTypeSelection

  const swapAsGrid = () => {
    props.setAsGrid(!asGrid)
  }
  const swapShowGenSelection = () => {
    props.setShowGenSelection(!showGenSelection)
  }
  const swapShowTypeSelection = () => {
    props.setShowTypeSelection(!showTypeSelection)
  }


  return (
      <div className={"topbar"}>
        <div>
          <input className={"search"} style={{width: 300}} placeholder={"SÃ¸k"}/>
          <Search style={{height: 35, width: 35, fill: "#F5F5F5"}}/>
        </div>
        <div>
          <Button variant="outline-light" style={{marginRight:5}} onClick={swapShowGenSelection}>Velg Generasjon</Button>
          <Button variant="outline-light" onClick={swapShowTypeSelection}>Velg Type</Button>
        </div>
        <div>
          <List style={{height: 35, width: 35}} className={asGrid ? "" : "active"} onClick={swapAsGrid}/>
          <Grid style={{height: 35, width: 35}} className={asGrid ? "active" : ""} onClick={swapAsGrid}/>
        </div>
      </div>
  );
}