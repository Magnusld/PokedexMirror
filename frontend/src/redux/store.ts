import { configureStore } from '@reduxjs/toolkit'
import selectedGenReducer from './generationSlice'
import selectedTypeReducer from './TypeSlice'

export const store = configureStore({
  reducer: {
    selectedGen: selectedGenReducer,
    selectedType: selectedTypeReducer
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch