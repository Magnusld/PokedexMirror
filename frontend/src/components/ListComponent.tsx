import React, {useEffect, useState} from 'react';
import '../style/App.css';
import {ListingComponent} from "./ListingComponent";
import {AfterInputFields, PokemonSimple, Variables, OrderByInputFields} from "../types";
import { useQuery } from '@apollo/client';
import {Button} from "react-bootstrap";
import gql from 'graphql-tag';
import { useSelector } from 'react-redux';
import { RootState } from '../redux/store';

export function ListComponent(props: {
  asGrid: boolean
}) {

  const [nextQueryIds, setNextQueryIds] = useState<any[]>([null, null])
  const [pageCounter, setPageCounter] = useState<number>(1)

  interface PokemonSimpleData {
    pokemons: PokemonSimple[]
  }


  const GET_POKEMON_DATA = gql`
  query($orderBy: [PokemonOrderByInput!], $where: PokemonWhereInput, $first: Int, $after: PokemonWhereUniqueInput, $last: Int, $before: PokemonWhereUniqueInput) {
    pokemons(orderBy: $orderBy, where: $where, first: $first, after: $after, last: $last, before: $before,) {
      id
      pokedexNr
      name
      generation
      type1
      type2
    }
  }
  `;

  const searchInput = useSelector((state: RootState) => state.searchInput.value)
  const selectedGen = useSelector((state: RootState) => state.selectedGen.value.filter(element => element.selected).map(element => element.id + 1))
  const selectedType = useSelector((state: RootState) => state.selectedType.value.filter(element => element.selected).map(element => element.name))
  const sorting = useSelector((state: RootState) => state.sort.value)

  /*
  useEffect(() => {
    console.log(searchInput)
    console.log(selectedGen)
    console.log(selectedType);
    console.log(sorting);
  }, [searchInput, selectedGen, selectedType, sorting])

   */


  /**
   * used by useQuery hook
   * @returns a set of variables to be used by graphQL query
   */
  function setQueryVariables() : object {
    let after = nextQueryIds[0]
    let before = nextQueryIds[1]
    let orderBy : OrderByInputFields = {}
    if(sorting.type === "name") {
      orderBy = {"name": sorting.ordering}
    } else if(sorting.type === "pokedexNr") {
      orderBy = {"pokedexNr": sorting.ordering}
    }
    let name : string | null = null
    searchInput ? name = searchInput : name = null
    //console.log("prevId i variablers: ", prev)
    //console.log("afterId i variablers: ", after)
    let first: number | null = 15
    let last: number | null = null
    if (pageCounter != 1){
      if (after == null && before != null){
        last = 15
        first = null
      }
    }
    //console.log("first: ",first,"last: ", last,"after: ", after, "before:", before)
    const variables : Variables = { 
      variables: {
        "orderBy" : orderBy,
        "first": first,
        "last": last,
        "after": after,
        "before": before,
        "where": {
          "type1": {
            "in": selectedType
          },
          "generation": {
            "in": selectedGen
          },
          "name": {
            "contains": name
          }
        }
      }
    }
    //console.log(variables)
    return variables;
  }

  const { loading, error, data } = useQuery<PokemonSimpleData, any>(GET_POKEMON_DATA, setQueryVariables());

  const asGrid = props.asGrid


  /**
   * used to render pokemon to grid/list
   * @returns a list with compontents
   */
  function showListing(pokeList: any[] | undefined) {
    if(pokeList != undefined) {
      return pokeList.map((pokemon : PokemonSimple, i) => (
        <ListingComponent key={i} asGrid={asGrid} pokemon={pokemon} />
      ))
    }
    else {
      return <div>Failed!</div>
    }
  }
  function showPrevListings() {
    setPageCounter(pageCounter - 1)
    //console.log(pageCounter)
    //console.log(data)
    if (pageCounter == 2) {
      setNextQueryIds([null, null])
    } else {
      if (data != undefined) {
        const newPrev: AfterInputFields = {
          id: data.pokemons[0].id
        }
        //console.log("prevId = ",newPrev)
        setNextQueryIds([null, newPrev])
      }
    }
  }
  function showFirstListings() {
    setPageCounter(1)
    setNextQueryIds([null, null])
  }


  function showNextListings() {
    setPageCounter(pageCounter + 1)
    //console.log(pageCounter)
    //console.log(data)
    if (data != undefined) {
      const newAfter: AfterInputFields = {
        id: data.pokemons[data.pokemons.length -1].id
      }
      //console.log("afterId = ",newAfter)
      setNextQueryIds([newAfter, null])
    }
  }

  function chooseClassName() {
    if (asGrid) {
      return "AsGrid"
    } else { return "AsList"}
  }

  if (error) return <div>Error! {error.message}</div>;

  return (
      <div>
          {loading ? (<p>Loading</p>) : (
            <div className={"list"+chooseClassName()}>
              {showListing(data?.pokemons)}
            </div>
          )}
          <div className={"showMore"}>
            <div className={"PageNavButtonGroups"}>
              <Button className={"showNextButton"} onClick={showFirstListings}>&lt;&lt; FÃ¸rste side</Button>
            </div>
            <div className={"PageNavButtonGroups"}>
              <Button className={"showNextButton"} onClick={showPrevListings} disabled={pageCounter==1} > &lt; Forrige side</Button>
              <div className="page-counter">{pageCounter}</div>
              <Button className={"showNextButton"} onClick={showNextListings} disabled={data?.pokemons.length != 15}>Neste side &gt; </Button>
            </div>
            <div className={"PageNavButtonGroups"}>
              {/*Funksjonalitet for en annen gang :)*/}
              <Button className={"showNextButton"} style={{visibility: "hidden"}}>Siste side &gt;&gt;</Button>
            </div>
          </div>
      </div>
  )
}