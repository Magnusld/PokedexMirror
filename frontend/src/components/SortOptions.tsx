import React, {useState} from 'react';
import '../style/App.css';
import {Button, Form, ToggleButton} from "react-bootstrap";
import {RootState} from "../redux/store";
import {useDispatch, useSelector} from "react-redux";
import {setAllGensFalse, swapSelectedGen} from "../redux/generationSlice";
import { changeSortOrder, changeSortType } from '../redux/sortSlice';

export function SortOptions() {
  const dispatch = useDispatch()

  /**
   * Change sorting type to select value
   * @param e event
   */
  function handleSortType(e: React.ChangeEvent<HTMLSelectElement>) {
      dispatch(changeSortType(e.target.value))
  }

  /**
   * Change sorting order to select value
   * @param e event
   */
  function handleSortOrder(e: React.ChangeEvent<HTMLSelectElement>) {
      dispatch(changeSortOrder(e.target.value))
  }

  return (
      <div style={{display: "flex", justifyContent: "center", width: "100%"}}>
        <div className={"filteringOptionsContainer sortOptions"}>
          <div>
            <h3 style={{marginTop: 10}}>Sortering</h3>
          </div>
          <div className="sort-form-select">
            <div className="sort-inner-select">
                <h4>Type</h4>
                <Form.Select aria-label="Sorting type" className="sort-select" onChange={handleSortType}>
                    <option value="pokedexNr" >Nummer</option>
                    <option value="name">Navn</option>
                </Form.Select>
            </div>
            <div className="sort-inner-select">
                <h4>Rekkef√∏lge</h4>
                <Form.Select aria-label="Sorting order" className="sort-select" onChange={handleSortOrder}>
                    <option value="asc">Stigende</option>
                    <option value="desc">Synkende</option>
                </Form.Select>
            </div>
            
          </div>
          
        </div>
      </div>
  )
}