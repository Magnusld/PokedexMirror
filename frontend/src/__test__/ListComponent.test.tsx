import * as ReactDOM from 'react-dom';
import React from "react";
import {ListComponent} from "../components/ListComponent"
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import { store } from '../redux/store'
import {ApolloProvider, ApolloClient, createHttpLink, InMemoryCache} from '@apollo/client';

const httpLink = createHttpLink({
    uri: 'http://localhost:4000'
});

const client = new ApolloClient({
    link: httpLink,
    cache: new InMemoryCache()
});


describe("Unit test on the ListComponent component", () => {
    let container: HTMLDivElement;
    beforeEach(() => {
        container = document.createElement("div");
        document.body.appendChild(container);

        ReactDOM.render(
            <ApolloProvider client={client}>
                <Provider store={store}>
                    <ListComponent asGrid={true} />
                </Provider>
            </ApolloProvider>, container)
    })

    afterEach(() => {
        document.body.removeChild(container);
        container.remove();
        jest.resetAllMocks();
    })

    it("matches snapshot for component", () => {
        const tree = renderer.create(
            <ApolloProvider client={client}>
                <Provider store={store}>
                    <ListComponent asGrid={true} />
                </Provider>
            </ApolloProvider>).toJSON()
        expect(tree).toMatchSnapshot()
    })

    it("renders listingComponents", ()=> {
        const numberOfListings = container.querySelectorAll("div").length
        expect(numberOfListings).toBeGreaterThan(0)
    })

})