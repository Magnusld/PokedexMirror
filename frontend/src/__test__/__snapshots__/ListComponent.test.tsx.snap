// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test on the ListComponent component matches snapshot for component 1`] = `
<div>
  <p>
    Loading
  </p>
  <div
    className="showMore"
  >
    <div
      className="PageNavButtonGroups"
    >
      <button
        className="showNextButton btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        &lt;&lt; FÃ¸rste side
      </button>
    </div>
    <div
      className="PageNavButtonGroups"
    >
      <button
        className="showNextButton btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
         &lt; Forrige side
      </button>
      <div
        className="page-counter"
      >
        1
      </div>
      <button
        className="showNextButton btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Neste side &gt; 
      </button>
    </div>
    <div
      className="PageNavButtonGroups"
    >
      <button
        className="showNextButton btn btn-primary"
        disabled={false}
        style={
          Object {
            "visibility": "hidden",
          }
        }
        type="button"
      >
        Siste side &gt;&gt;
      </button>
    </div>
  </div>
</div>
`;
